### ISSUE TITLE
Ability to load images from files

### ISSUE THREAD

**Issue by QuantumBadger** [QuantumBadger/Speedy2D#1]
As suggested on Reddit, it would be handy to be able to load images directly from either a `u8` slice, or a filename.

**Comment by MoAlyousef** [comment]
Just a few days ago I was looking for such a crate. Trying to make speedy2D work rendering an image (using the image crate) gives me several errors in the glwrapper. Example code (maybe I'm using speedy2D wrong):
```rust
extern crate image;
use image::GenericImageView;

use speedy2d::color::Color;
use speedy2d::dimen::Vector2;
use speedy2d::image::{ImageDataType, ImageSmoothingMode};
use speedy2d::window::{WindowHandler, WindowHelper};
use speedy2d::{Graphics2D, Window};

fn main() {
    simple_logger::SimpleLogger::new().init().unwrap();
    let window = Window::new_centered("Speedy2D: Hello World", (640, 240)).unwrap();
    let image = image::open("test/assets/expected_images/test_half_circle.png").unwrap();
    window.run_loop(MyWindowHandler { image })
}

struct MyWindowHandler {
    image: image::DynamicImage,
}

impl WindowHandler for MyWindowHandler {
    fn on_draw(&mut self, helper: &mut WindowHelper, graphics: &mut Graphics2D) {
        graphics.clear_screen(Color::WHITE);
        let size = self.image.dimensions();
        let _handle = graphics
            .create_image_from_raw_pixels(
                ImageDataType::RGBA,
                ImageSmoothingMode::Linear,
                Vector2::new(size.0, size.1),
                &self.image.to_rgba8(),
            )
            .unwrap();
        helper.request_redraw();
    }
}
```

The error I get (along with the logs and backtrace):
```
2021-02-28 19:39:39,548 TRACE [mio::poll] registering with poller
2021-02-28 19:39:39,548 TRACE [mio::poll] registering with poller
2021-02-28 19:39:39,548 TRACE [mio::poll] registering with poller
2021-02-28 19:39:39,548 TRACE [mio::poll] registering with poller
2021-02-28 19:39:39,548 TRACE [mio::poll] registering with poller
2021-02-28 19:39:39,550 DEBUG [speedy2d::window] Monitor #0 (primary): default
2021-02-28 19:39:39,551 INFO  [speedy2d::window] Trying vsync=true, multisampling=8...
2021-02-28 19:39:39,561 INFO  [winit::platform_impl::platform::x11::window] Guessed window scale factor: 1
2021-02-28 19:39:39,561 DEBUG [winit::platform_impl::platform::x11::window] Calculated physical dimensions: 640x240
2021-02-28 19:39:39,707 INFO  [speedy2d::window] Failed with error: CreationErrors([NoAvailablePixelFormat, OsError("GL context creation failed")])
2021-02-28 19:39:39,707 INFO  [speedy2d::window] Trying vsync=true, multisampling=4...
2021-02-28 19:39:39,709 INFO  [winit::platform_impl::platform::x11::window] Guessed window scale factor: 1
2021-02-28 19:39:39,709 DEBUG [winit::platform_impl::platform::x11::window] Calculated physical dimensions: 640x240
2021-02-28 19:39:39,736 INFO  [speedy2d::window] Failed with error: CreationErrors([NoAvailablePixelFormat, OsError("GL context creation failed")])
2021-02-28 19:39:39,736 INFO  [speedy2d::window] Trying vsync=true, multisampling=2...
2021-02-28 19:39:39,755 INFO  [winit::platform_impl::platform::x11::window] Guessed window scale factor: 1
2021-02-28 19:39:39,756 DEBUG [winit::platform_impl::platform::x11::window] Calculated physical dimensions: 640x240
2021-02-28 19:39:39,813 INFO  [speedy2d::window] Failed with error: CreationErrors([NoAvailablePixelFormat, OsError("GL context creation failed")])
2021-02-28 19:39:39,813 INFO  [speedy2d::window] Trying vsync=true, multisampling=1...
2021-02-28 19:39:39,856 INFO  [winit::platform_impl::platform::x11::window] Guessed window scale factor: 1
2021-02-28 19:39:39,856 DEBUG [winit::platform_impl::platform::x11::window] Calculated physical dimensions: 640x240
2021-02-28 19:39:39,900 INFO  [winit::platform_impl::platform::x11::window] Guessed window scale factor: 1
2021-02-28 19:39:39,900 DEBUG [winit::platform_impl::platform::x11::window] Calculated physical dimensions: 640x240
2021-02-28 19:39:39,938 INFO  [speedy2d::window] Context created
2021-02-28 19:39:39,944 INFO  [speedy2d::window] Centering window. Monitor size: PhysicalSize { width: 1536, height: 864 }. Window outer size: PhysicalSize { width: 640, height: 240 }.
2021-02-28 19:39:39,947 INFO  [speedy2d::window] Using OpenGL version: 3.1 Mesa 20.0.8
2021-02-28 19:39:39,947 INFO  [speedy2d::window] Setting up GL debug log
2021-02-28 19:39:39,949 INFO  [speedy2d::glwrapper] GL context manager created
2021-02-28 19:39:40,062 INFO  [speedy2d::window] Resized: PhysicalSize { width: 640, height: 240 }
thread 'main' panicked at 'already mutably borrowed: BorrowError', src/glwrapper.rs:866:11
stack backtrace:
   0: rust_begin_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:483
   1: core::panicking::panic_fmt
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/panicking.rs:85
   2: core::option::expect_none_failed
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/option.rs:1234
   3: core::result::Result<T,E>::expect
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:933
   4: core::cell::RefCell<T>::borrow
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/cell.rs:792
   5: speedy2d::glwrapper::is_gl_context_valid
             at ./src/glwrapper.rs:866
   6: speedy2d::glwrapper::is_gl_context_valid_weak
             at ./src/glwrapper.rs:875
   7: speedy2d::glwrapper::GLHandle::is_context_still_valid
             at ./src/glwrapper.rs:225
   8: <speedy2d::glwrapper::GLHandle as core::ops::drop::Drop>::drop
             at ./src/glwrapper.rs:233
   9: core::ptr::drop_in_place
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:175
  10: core::ptr::drop_in_place
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:175
  11: <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/rc.rs:1207
  12: core::ptr::drop_in_place
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:175
  13: core::ptr::drop_in_place
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:175
  14: speedy2d::glwrapper::GLContextManager::bind_texture
             at ./src/glwrapper.rs:992
  15: speedy2d::glwrapper::GLTexture::set_image_data
             at ./src/glwrapper.rs:786
  16: speedy2d::renderer2d::Renderer2D::create_image_from_raw_pixels
             at ./src/renderer2d.rs:620
  17: speedy2d::Graphics2D::create_image_from_raw_pixels
             at ./src/lib.rs:445
  18: <image::MyWindowHandler as speedy2d::window::WindowHandler>::on_draw
             at ./examples/image.rs:25
  19: <speedy2d::window::DrawingWindowHandler<Handler,UserEventType> as speedy2d::window::WindowImplHandler<UserEventType>>::on_draw::{{closure}}
             at ./src/window.rs:917
  20: speedy2d::GLRenderer::draw_frame
             at ./src/lib.rs:402
  21: <speedy2d::window::DrawingWindowHandler<Handler,UserEventType> as speedy2d::window::WindowImplHandler<UserEventType>>::on_draw
             at ./src/window.rs:917
  22: speedy2d::window::WindowImpl<UserEventType>::loop_iter
             at ./src/window.rs:517
  23: speedy2d::window::WindowImpl<UserEventType>::run_loop::{{closure}}
             at ./src/window.rs:554
  24: winit::platform_impl::platform::sticky_exit_callback
             at /home/rayloom/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/mod.rs:736
  25: winit::platform_impl::platform::x11::EventLoop<T>::run_return
             at /home/rayloom/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/x11/mod.rs:320
  26: winit::platform_impl::platform::x11::EventLoop<T>::run
             at /home/rayloom/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/x11/mod.rs:398
  27: winit::platform_impl::platform::EventLoop<T>::run
             at /home/rayloom/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/mod.rs:652
  28: winit::event_loop::EventLoop<T>::run
             at /home/rayloom/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/event_loop.rs:154
  29: speedy2d::window::WindowImpl<UserEventType>::run_loop
             at ./src/window.rs:546
  30: speedy2d::Window<UserEventType>::run_loop
             at ./src/lib.rs:975
  31: image::main
             at ./examples/image.rs:14
  32: core::ops::function::FnOnce::call_once
             at /home/rayloom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
2021-02-28 19:39:40,160 INFO  [speedy2d::glwrapper] GL context manager is now inactive
```

**Comment by QuantumBadger** [comment]
@MoAlyousef Thanks for the report! You're using it correctly, this looks like an internal issue. Just looking into it now.

**Comment by QuantumBadger** [comment]
@MoAlyousef The issue is now fixed, it was because the texture from the previous frame was being deleted at the same time the new texture was being bound. I've added an automated test for this so it won't occur again.

The fix is in v1.0.1, which I've published. If you run into any more issues, please do let me know!

**Comment by MoAlyousef** [comment]
Thanks. I can confirm. I'm running into another problem for which I'll open a new issue.
